### ISSUE TITLE
[RFC]: simplify model stage config for end users

### ISSUE THREAD

**Issue by ZJY0516** [vllm-project/vllm-omni#74]
### Motivation.

Currently we use yaml to configure almost everything. However, it is **too complicated** and **does not clearly distinguish between user configurations and model developer configurations**.
For example, `worker_cls` and `model_arch` should not be exposed to end users who only wish to customize runtime behavior.

### Proposed Change.

We should define immutable config in model definition files like
```python
class Model(ComposedStageBase):

    def create_stages(self, server_args: ServerArgs):
        """Set up stages with proper dependency injection."""

        self.add_stage(
            stage_name="input_validation_stage", stage=InputValidationStage()
        )

        self.add_stage(
            stage_name="thinker", stage=Thinker()
        )

        self.add_stage(
            stage_name="talker", stage=Talker()
        )
```

As for configurations that are meant to be overridden by users, we can continue to keep them in a YAML configuration file.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by ZJY0516** [comment]
Config below should not be exposed to end user
```
      model_stage: thinker
      model_arch: Qwen2_5OmniForConditionalGeneration
      worker_cls: vllm_omni.worker.gpu_ar_worker.GPUARWorker
      scheduler_cls: vllm_omni.core.sched.scheduler.OmniScheduler

    custom_process_input_func: vllm_omni.model_executor.stage_input_processors.qwen2_5_omni.thinker2talker

  edges:
    - from: 0                   # thinker → talker: trigger only after receiving full input (-1)
      to: 1
      window_size: -1
    - from: 1                   # talker → code2wav: trigger only after receiving full input (-1)
      to: 2
      window_size: -1
```

**Comment by Gaohan123** [comment]
Thanks! The optimization direction is definitely important. Welcome example in more details referring to a specific model.

**Comment by yinpeiqi** [comment]
Hi all, I have propose a more detailed version for the model stage abstraction. @ZJY0516 @Gaohan123 

## Motivation

Currently, its hard to implement new model, and the frontend of the `examples/qwen2.5omni` is hard to implement. If the user want to implement a model in `transformers` , the users will write:

```python
class Thinker(LLM):
	def forward():
		# implement forward for one step

class Talker(LLM):
	def forward():
		# implement forward for one step

class Token2Wav(DiT):
	def forward()

class Qwen2_5Omni():
	def generate():
		# 1. process inputs
		...
		# 2. execute thinker
		thinker_result = self.thinker.generate(...)
		# 3. process thinker to talker
		...
		# 4. executor talker
		talker_result = self.talker.generate(...)
		# 5. talker to token2wav
		...
		# 6. token2wav
		wav_result = self.token2wav.chunked_decode(....)
		return ...
```

Under such `generate` process, the user only need to consider data transformations and transitions between different LLM/DiT stages. While each LLM stage only need to consider one step, same as the process of one single LLM model. This implement is simple.

However, implementing a new Omni model in vllm-omni is challenging. The existing interface (as seen in `examples/qwen2.5omni`) tightly couples these two distinct responsibilities within a single forward function (i.e., stage management and step execution).

Current implement in `example`  folder:

```python
# for the classes Thinker, Talker, and Token2Wav, the 
# implementation is similar to Transformers, the user only
# need to consider one step of the LLM.

class Qwen2_5Omni():
	def forward():
		# one step for thinker
		if self.stage == "thinker":
			...
		# one step for talker
		if self.stage == "talker":
			if prefill:
				...
			else:
				...
		# one step for token2wav
		if self.stage == "token2wav":
			...
```

This abstraction forces users to manually orchestrate the complex state management between stages, making the code difficult to write, maintain, and customize. While the standard transformers generate approach simplifies this by handling one step at a time (inefficient for our needs), our goal of packing stages into a unified vLLM Engine for performance has inadvertently created a high barrier to entry for new model integration.

## **Proposed Change**

I agreed with @ZJY0516  for the stage execution (follows sglang diffusion), while with the below differences. The core principle is to decouple the connection topology (how models are chained) from the execution details (how a model performs a step).

1. We need to follow to principle of `transformers` , which the execution pipeline of a Omni model should focus on the connection between several base LLM/DiT models. The Pipeline don’t need to consider what’s happening inside an AR or DiT model (what happening in each step), it only need to know the whole pipeline is enough. The execution detail of the pipeline should be implement internally by our framework (e.g., what to process for prefill stage and decode stage).
 The whole pipeline should look likes:

`(AR)Thinker` - `processFN` - `(AR)Talker`- `processFN` - `(DiT)Token2Wav`

2. The execution of each pipeline stage (AR or DiT) need to include a barrier (also like window size) to control whether the processing could be forward to the following stages. For example, if the barrier is true, that means the next stage can process only when the current stage is finished. If not, that means the next stage could receive partial output for following process. Two examples are:
    1. return text-only for thinker output
    2. streaming output for Code2Wav (e.g., Qwen3Omni, LongCat-Flash-Omni)
 
3. The stage infos should not fixed by the model. For example, when serving Omni model, the server may receive both 1. text-only and 2. return audio requests, in this case the stage should be fixed.

<img width="1617" height="788" alt="Image" src="https://github.com/user-attachments/assets/d79087d6-7404-4b07-9134-172f84efcc17" />

I propose an example for the abstraction:

```python

class Qwen2_5Omni(PipelineBase):
	def __init__(self, yamlfile):
		self.pipeline = [
				ARStage(model=self.thinker,
								process_fn=self.process_input,
								barrier=True), # encoder inside it
				ARStage(model=self.talker, 
								process_fn=self.thinker_to_talker,
								barrier=True),
				DiTStage(model=self.Token2Wav, 
								 process_fn=self.talker_to_token2wav,
								 barrier=True),
		]
		# extend from the PipelineBase. PipelineBase is handled by the Orchestrator
		super().construct(self.pipeline, yamlfile)
	
	def thinker_to_talker(**kwargs):
		...
	
	def talker_to_token2wav(**kwargs):
		...
	# we'd better don't let user write a detail forward here.
	def forward(..., bool text_only):
			self.pipeline.forward(**kwargs)
```



For the front-end, the user only need to customize the stages and the preprocess function.

For the back-end, the logits should be the same as the current code.

```python
class PipelineBase:
	def forward(reqs):
		stage_model = stage_dict[reqs.stage]
		if not reqs.processed:
			stage_model.run_process_fn(reqs)
		stage_model.forward(reqs)
		return ...
```

Here inside pipelineBase, we 1. execute the process function, and 2. process one step for LLM/DiT.

Basic Design Plan

1. Support the abstraction with PipelineBase
2. Add a Stage Manager in the Orchestrator, allow for window output
3. Bing the stage configs to a single requests, such that a text-only req and a audio req could be batch (i.e., batch in the thinker model), and manage their output independently.

**Comment by hsliuustc0106** [comment]
@qibaoyuan I think this RFC will help MiMo Audio
